Port to qhull_r

jspricke opened the pull request.

jspricke committed:
Port to qhull_r

jspricke commented:
Unit tests started to segfault when integrating 0.7.0 into Debian. As the non reentrant qhull version is deprecated anyhow, I ported it to qhull_r.

v4hn commented on src/bodies.cpp:
```diff
@@ -933,20 +913,24 @@ void bodies::ConvexMesh::useDimensions(const shapes::Shape* shape)
   char flags[] = "qhull Tv Qt";
-  int exitcode = qh_new_qhull(3, mesh->vertex_count, points, true, flags, null, null);
+  qhT qh_qh;
+  qhT *qh= &qh_qh;
```
```suggestion
  qhT* qh = &qh_qh;
```

v4hn reviewed (commented):
This change apparently breaks support with ROS kinetic. I guess `libqhull_r` was not available in Ubuntu 16.04?

Sounds like it's time to move the minimum supported distribution for the MoveIt stack to melodic.

Please have a look at CI, the melodic builds fail because of missing function symbols:
https://travis-ci.com/github/ros-planning/geometric_shapes/jobs/345628627#L615

jspricke commented:
>This change apparently breaks support with ROS kinetic. I guess `libqhull_r` was not available in Ubuntu 16.04?

Yeah, looks like I only added it in 2017:

http://changelogs.ubuntu.com/changelogs/pool/universe/q/qhull/qhull_2015.2-4/changelog

>Sounds like it's time to move the minimum supported distribution for the MoveIt stack to melodic.

Already, I moved all robots to Melodic like half a year ago ;). Will you 
it?

>Please have a look at CI, the melodic builds fail because of missing function symbols:
>https://travis-ci.com/github/ros-planning/geometric_shapes/jobs/345628627#L615

Will have a look tonight.

>> @@ -933,20 +913,24 @@ void bodies::ConvexMesh::useDimensions(const shapes::Shape* shape)
>   static FILE* null = fopen("/dev/null", "w");
>
>   char flags[] = "qhull Tv Qt";
>-  int exitcode = qh_new_qhull(3, mesh->vertex_count, points, true, flags, null, null);
>+  qhT qh_qh;
>+  qhT *qh= &qh_qh;
>
>```suggestion
>  qhT* qh = &qh_qh;
>```

Oh yeah, c&p from Qhull, fixed now.

jspricke commented:
>Please have a look at CI, the melodic builds fail because of missing function symbols:
>https://travis-ci.com/github/ros-planning/geometric_shapes/jobs/345628627#L615

Fixed.

v4hn reviewed (approved):
I approve this change. qhull is a pain in the a** and using the reentrant interface makes sense for `geometric_shapes`.

Please note that this breaks with Ubuntu 16.04, effectively breaking support of the whole MoveIt `master` stack in ROS kinetic. Before merge, this should be discussed at the maintainer meeting.
A few other changes can also be done at the same time (c++17 dependency, `boost::python`, ?), but we still had ROS kinetic users building `master` a while back.
If we merge this, CI setups for `kinetic` have to be removed in multiple repositories.

v4hn commented:
@tylerjw @rhaschke I know you already released this package into noetic, but I would propose to get this patch in before the main MoveIt release and re-release geometric_shapes along with it.

Strictly speaking, this is not necessary for noetic compatibility, but rather is the first change that would be relevant "beyond noetic".
Also it is a good patch to occasion disabling kinetic builds in CI :-)

rhaschke committed:
Remove unused qhull_header variable

rhaschke commented on cmake/FindQhull.cmake:
```diff
@@ -12,30 +12,29 @@ set(QHULL_DEBUG_NAME qhull_rd)
 set(QHULL_HEADER "${QHULL_HEADER}" CACHE INTERNAL "QHull header" FORCE )
 
 if(QHULL_HEADER)
-  get_filename_component(qhull_header ${QHULL_HEADER} NAME_WE)
```
As far as I can see, this `qhull_header` variable is not used.

rhaschke reviewed (commented):
I added a commit to remove an unused cmake variable (by editor also cleaned up trailing whitespace).
I updated .travis.yml to build on Noetic + Melodic only.

rhaschke commented:
Attempting to fix the Travis issues, I messed up this PR even more. Sorry for that.
I suggest reverting my last commit (`Updating .travis.yml` - only @jspricke can do this) and merge this as is.
To fix Travis (and clang-format-10), I have prepared new commits running on [Travis right now](https://travis-ci.com/github/rhaschke/geometric_shapes/builds/182073116).

jspricke commented:
@rhaschke removed your commit

rhaschke commented:
Thanks. Merging as is - with broken Travis, which will get fixed via another PR.
